openwhisk/executor.go

```
@@ -68,17 +95,160 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+			format:  FormatDatadog,
+		}
 	}
+	return nil
+}
+
+// consumeStream consumes a log stream into the given channel.
+func consumeStream(streamName string, stream io.Reader, ch chan LogLine) error {
+	s := bufio.NewScanner(stream)
+	for s.Scan() {
+		ch <- LogLine{
+			Time:    time.Now(),
+			Message: s.Text(),
+			Stream:  streamName,
+		}
+	}
+	return s.Err()
 }
 
 // Interact interacts with the underlying process
 func (proc *Executor) Interact(in []byte) ([]byte, error) {
-	// input to the subprocess
+	if proc.lines == nil {
+		// Remote logging is not configured. Just roundtrip and return immediately.
+		return proc.roundtrip(in)
+	}
+
+	// Fetch metadata from the incoming parameters
+	var metadata activationMetadata
+	if err := json.Unmarshal(in, &metadata); err != nil {
+		return nil, fmt.Errorf("failed to parse activation metadata: %w", err)
+	}
+
+	var grp errgroup.Group
+	grp.Go(func() error {
+		var sawStdoutSentinel, sawStdErrSentinel bool
+		var lastErr error
+		for line := range proc.lines {
+			if line.Message == logSentinel {
+				if line.Stream == "stdout" {
+					sawStdoutSentinel = true
+				} else {
+					sawStdErrSentinel = true
+				}
+				if sawStdoutSentinel && sawStdErrSentinel {
+					// We've seen both sentinels. Stop consuming until the next request.
+					return lastErr
+				}
+				continue
+			}
+
+			line.ActivationId = metadata.ActivationId
+			line.ActionName = metadata.ActionName
+			if err := proc.remoteLogger.Send(line); err != nil {
+				// We want to continue consuming all of the logs so we don't exit immediately
+				// but surface the error eventually.
+				lastErr = fmt.Errorf("failed to send log to remote location: %w", err)
+			}
+		}
+		return lastErr
+	})
+
+	proc.logout.WriteString(remoteLogNotice)
+	proc.logout.WriteString("\n")
+	out, err := proc.roundtrip(in)
+
+	// Wait for the streams to process completely.
+	if err := grp.Wait(); err != nil {
+		fmt.Fprintf(proc.logerr, "Failed to process logs: %v\n", err)
```

its probably good to return from here rather than continuing through the rest of the statements below

---

Hmm, I see where you're going but we have to write the OutputGuards actually. Maybe I should put them into a `defer` and then exit early here.

---

I have to correct: We actually don't want to return this error but the error from roundtrip everywhere. I'll do the defer and delete the flush, that'll come down to roughly the same thing then.